File: LC/python/Subsets.py
class Solution:
    # @param S, a list of integer
    # @return a list of lists of integer
    def subsets(self, S):
        def dfs(sub, start, depth):
            res.append(sub)
            if depth == len(S):
                return
            for i in range(start,len(S)):
                dfs(sub+[S[i]],i+1,depth+1)
        S.sort()
        res = []
        dfs([],0,0)
        return res
             
        

File: LC/java/Subsets.java
public class Solution {
    public ArrayList<ArrayList<Integer>> subsets(int[] S) {
        ArrayList<ArrayList<Integer>> res = new ArrayList<ArrayList<Integer>>();
        ArrayList<Integer> sub = new ArrayList<Integer>();
        
        Arrays.sort(S);
        dfs(res,sub,S,0);
        return res;
    }
    
    public void dfs(ArrayList<ArrayList<Integer>> res,ArrayList<Integer> sub,int[] S,int start){
        res.add(new ArrayList<Integer>(sub));
        
        for(int i = start; i<S.length; i++){
            sub.add(S[i]);
            dfs(res,sub,S,i+1);
            sub.remove(sub.size()-1);
        }
    }
}